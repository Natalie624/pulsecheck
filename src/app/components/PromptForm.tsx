// Prompt form

'use client'

import { useState, useRef } from 'react'
import ReactMarkdown from 'react-markdown'

export default function DashboardPage() {
  const [prompt, setPrompt] = useState('')
  const [tone, setTone] = useState('friendly') // default tone
  const [section, setSection] = useState('wins')
  const [team, setTeam] = useState('')
  const [timeframe, setTimeframe] = useState('')
  const [error, setError] = useState('')
  const [summary, setSummary] = useState('')
  const [isGenerating, setIsGenerating] = useState(false)

  const isSubmitting = useRef(false) // For debounce

  const CHARACTER_LIMIT = 1000

  const handleGenerate = async () => {
    // Prevent double-submit
    if (isSubmitting.current || isGenerating) return

    if (!prompt.trim()) {
      setError('Please enter a prompt.')
      return
    }

    if (prompt.length > CHARACTER_LIMIT) {
      setError(`Prompt exceeds the ${CHARACTER_LIMIT}-character limit.`)
      return
    }

    setError('')
    setIsGenerating(true)
    isSubmitting.current = true
    setSummary('') // clear previous summary

    // API call
    try {
      const response = await fetch('/api/generate', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ 
          prompt, 
          tone, 
          section,
          team,
          timeframe, 
        }),
      })

      const data = await response.json()

      if (!response.ok) {
        console.error('API Error:', data?.error)
        setError('Something went wrong. Please try again.')
      } else {
        setSummary(data.result.text)
      }
    } catch (err) {
      console.error('Unexpected Error:', err)
      setError('Unexpected error occurred. Please try again.')
    } finally {
      setIsGenerating(false)
      isSubmitting.current = false
    }
  }

  const handlePromptChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const input = e.target.value
    const trimmed = input.slice(0, CHARACTER_LIMIT)
    setPrompt(trimmed)

        if (input.length > CHARACTER_LIMIT) {
            setError(`Prompt exceeds the ${CHARACTER_LIMIT}-character limit.`)
        } else if (error && input.trim()) {
            setError('')
        }
    }

  return (
    <div className="max-w-2xl mx-auto py-10 px-4 text-gray-800 text-left w-full">
        <div className="flex justify-center">
            <textarea
                className="w-full md:w-[40rem] h-32 p-3 border border-gray-300 rounded mb-2"
                placeholder="Enter your prompt here..."
                value={prompt}
                onChange={handlePromptChange}
            />
        </div>
      

      <div className="flex justify-between items-center text-sm mb-2">
        <span className="text-gray-400">
          Character limit: {prompt.length} / {CHARACTER_LIMIT}
        </span>
        {prompt.length > CHARACTER_LIMIT && (
          <span className="text-red-600 font-medium" role="alert">
            Limit exceeded
          </span>
        )}
      </div> 

      {error && (
        <p className="text-red-600 text-sm mb-2" role="alert">
          {error}
        </p>
      )}

      {/* Section Dropdown */}
      <select
        className="w-full p-3 border border-gray-300 rounded mb-4"
        value={section}
        onChange={(e) => setSection(e.target.value)}
      >
        <option value="wins">Wins</option>
        <option value="risks">Risks</option>
        <option value="blockers">Blockers</option>
        <option value="dependencies">Dependencies</option>
        <option value="nextSteps">Next Steps</option>
      </select>


      {/* Tone Dropdown */}
      <select
        className="w-full p-3 border border-gray-300 rounded mb-4"
        value={tone}
        onChange={(e) => setTone(e.target.value)}
      >
        <option value="friendly">Team Chill</option>
        <option value="formal">Executive Ready</option>
        <option value="urgent">Escalation mode</option>
      </select>

      {/* Optional Team Name Input */}
      <input
        type="text"
        placeholder="Team Name (optional)"
        className="w-full p-3 border border-gray-300 rounded mb-4"
        value={team}
        onChange={(e) => setTeam(e.target.value)}
      />

      {/* Optional Timeframe Input */}
      <input
        type="text"
        placeholder="Timeframe (ie; this week, past sprint)"
        className="w-full p-3 border border-gray-300 rounded mb-4"
        value={timeframe}
        onChange={(e) => setTimeframe(e.target.value)}
      />

      <div className="flex justify-center">
        <button
            className="w-full md:w-auto bg-[#6c47ff] text-white hover:bg-[#5a38e0] rounded-full font-medium text-sm sm:text-base h-10 sm:h-12 px-6 cursor-pointer transition disabled:opacity-50"
            onClick={handleGenerate}
            disabled={isGenerating}
        >
            {isGenerating ? 'Generating...' : 'Generate'}
        </button>
      </div>

      {summary && (
        <div className="mt-6 p-4 bg-gray-100 border border-gray-300 rounded">
          <h2 className="font-semibold mb-2 text-gray-700">Generated Summary:</h2>
          <div className="prose pros-sm sm:prose-base max-w-none text-gray-800">
            <ReactMarkdown>{summary}</ReactMarkdown>
          </div>
        </div>
      )}
    </div>
  )
}