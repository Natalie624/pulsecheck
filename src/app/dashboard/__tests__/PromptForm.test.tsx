// src/components/__tests__/PromptForm.test.tsx

import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import PromptForm from '../../components/PromptForm';
import '@testing-library/jest-dom'
import { mockFetchResponse } from './testUtils/mockFetch';

vi.mock('react-markdown', () => ({
  __esModule: true,
  default: ({ children }: { children: string }) => <div>{children}</div>,
}))

describe('PromptForm', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
  })

  it('renders textarea, dropdowns, and button', () => {
    render(<PromptForm />)
    expect(screen.getByPlaceholderText(/enter your prompt/i)).toBeInTheDocument()
    expect(screen.getByRole('combobox', { name: /section/i })).toBeInTheDocument()
    expect(screen.getByRole('combobox', { name: /tone/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /generate/i })).toBeInTheDocument()
  })

  it('shows an error if prompt is empty on submit', async () => {
    render(<PromptForm />)
    fireEvent.click(screen.getByRole('button', { name: /generate/i }))
    expect(await screen.findByText(/please enter a prompt/i)).toBeInTheDocument()
  })

  it('truncates prompt over character limit and shows error', async () => {
    render(<PromptForm />)
    const textarea = screen.getByPlaceholderText(/enter your prompt/i)
    fireEvent.change(textarea, { target: { value: 'a'.repeat(2000) } })
    expect(textarea).toHaveValue('a'.repeat(1000))
    expect(screen.getByText(/prompt exceeds/i)).toBeInTheDocument()
  })

  it('submits prompt and displays summary from API', async () => {
    const mockSummary = '✅ Key results achieved'
    mockFetchResponse({ result: { text: mockSummary } })

    render(<PromptForm />)
  fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), { 
    target: { value: 'Sprint wins' },
  })
  fireEvent.click(screen.getByRole('button', { name: /generate/i }))

  await waitFor(() => {
    expect(screen.getByText(/generated summary/i)).toBeInTheDocument()
    expect(screen.getByText(mockSummary)).toBeInTheDocument()
  })
  })

  it('shows an error if API returns an error', async () => {
    mockFetchResponse({ error: 'Bad Request' }, false)

    render(<PromptForm />)
    fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), {
      target: { value: 'Sprint update' },
    })

    fireEvent.click(screen.getByRole('button', { name: /generate/i }))

    expect(await screen.findByText(/something went wrong/i)).toBeInTheDocument()
  })

  it('disables and re-enables the button while generating', async () => {
    mockFetchResponse({ result: { text: '✅ Summary' } })

    render(<PromptForm />)
    fireEvent.change(screen.getByPlaceholderText(/enter your prompt/i), {
      target: { value: 'Testing loading state' },
    })

    const button = screen.getByRole('button', { name: /generate/i })
    fireEvent.click(button)
    expect(button).toBeDisabled()

    await waitFor(() => expect(button).not.toBeDisabled())
  })

  it('matches snapshot', () => {
  const { container } = render(<PromptForm />)
  expect(container).toMatchSnapshot()
})

})
