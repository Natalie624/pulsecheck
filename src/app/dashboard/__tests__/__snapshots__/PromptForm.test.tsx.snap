// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PromptForm > matches snapshot 1`] = `
<div>
  <div
    class="max-w-2xl mx-auto py-10 px-4 text-gray-800 text-left w-full"
  >
    <div
      class="flex justify-center"
    >
      <textarea
        class="w-full md:w-[40rem] h-32 p-3 border border-gray-300 rounded mb-2"
        placeholder="Enter your prompt here..."
      />
    </div>
    <div
      class="flex justify-between items-center text-sm mb-2"
    >
      <span
        class="text-gray-400"
      >
        Character limit: 
        0
         / 
        10000
      </span>
    </div>
    <label
      class="flex items-center text-sm font-medium text-gray-700 mb-1"
    >
      Section
      <button
        aria-label="More info"
        class="ml-1 text-gray-400 hover:text-gray-700 focus:outline-none"
        data-state="closed"
      >
        ℹ️
      </button>
    </label>
    <select
      aria-label="section"
      class="w-full p-3 border border-gray-300 rounded mb-4"
    >
      <option
        value="wins"
      >
        Wins
      </option>
      <option
        value="risks"
      >
        Risks
      </option>
      <option
        value="blockers"
      >
        Blockers
      </option>
      <option
        value="dependencies"
      >
        Dependencies
      </option>
      <option
        value="nextSteps"
      >
        Next Steps
      </option>
    </select>
    <label
      class="flex items-center text-sm font-medium text-gray-700 mb-1"
    >
      Tone
      <button
        aria-label="More info"
        class="ml-1 text-gray-400 hover:text-gray-700 focus:outline-none"
        data-state="closed"
      >
        ℹ️
      </button>
    </label>
    <select
      aria-label="tone"
      class="w-full p-3 border border-gray-300 rounded mb-4"
    >
      <option
        value="friendly"
      >
        Team Chill
      </option>
      <option
        value="formal"
      >
        Executive Ready
      </option>
      <option
        value="urgent"
      >
        Escalation mode
      </option>
    </select>
    <input
      class="w-full p-3 border border-gray-300 rounded mb-4"
      placeholder="Team Name (optional)"
      type="text"
      value=""
    />
    <input
      class="w-full p-3 border border-gray-300 rounded mb-4"
      placeholder="Timeframe (ie; this week, past sprint)"
      type="text"
      value=""
    />
    <div
      class="flex justify-center"
    >
      <button
        class="w-full md:w-auto bg-[#6c47ff] text-white hover:bg-[#5a38e0] rounded-full font-medium text-sm sm:text-base h-10 sm:h-12 px-6 cursor-pointer transition disabled:opacity-50"
      >
        Generate
      </button>
    </div>
  </div>
</div>
`;
